### ISSUE TITLE
flycheck looking in wrong path

### ISSUE THREAD

**Issue by arnyeinstein** [flycheck/flycheck#1374]
Hi
When I activate flycheck-mode on a R-file buffer, an error is reported:

```
Suspicious state from syntax checker r-lintr: Flycheck checker r-lintr returned non-zero exit code 1, but its output contained no errors:
'c:\PROGRA~1\R\R-34~1.3/bin/x64/Rterm.exe" --slave --restore --no-save -e "library' 
is not recognized as an internal or external command, operable program or batch file.
```

In my .emacs file, I have the following line:

> (setq inferior-R-program-name "C:/Program Files/R/R-3.4.3/bin/x64/Rterm.exe")

Here the output from verifying:

> Syntax checkers for buffer unif.R in ess-mode:
> 
>   r-lintr
>     - may enable:    yes
>     - predicate:     t
>     - executable:    Found at C:/Program Files/R/R-3.4.3/bin/x64/R.exe
>     - lintr library: present

> --------------------
> 
> Flycheck version: 32snapshot (package: 20171214.1215)
> Emacs version:    25.3.1
> System:           x86_64-w64-mingw32
> Window system:    w32

I have no idea why the syntax in R-file is not working

THanks
Renger

**Comment by cpitclaudel** [comment]
Thanks for the report.  There seems to be multiple problems with the R checker, starting with the fact that it breaks `flycheck-verify-setup` (cc @marsam in 43569c9321684d3bf3be9ef9d3718c151d28bc13 — that code breaks when R isn't installed).

I don't have a Windows system easily available to reproduce your problem.  Can you show what happens when you run `C-c ! C-c r-lintr RET`?

**Comment by arnyeinstein** [comment]
Then I get this:


 "C:/Program` Files/R/R-3.4.3/bin/x64/R.exe" "--slave" "--restore" "--no-save" "-e" 
  library(lintr);try(lint(commandArgs(TRUE), cache=TRUE, default_linters))" "--args" "d:/Projects/2017_CAN/Model/R/unif.R"
 'C:\PROGRA~1\R\R-34~1.3/bin/x64/Rterm.exe" --slave --restore --no-save -e "library' is not recognized as an internal or external command, operable program or batch file.

Renger

**Comment by cpitclaudel** [comment]
> (I had to add empty spaces before and after the ~-sign, to get a correct view in the post)

That still makes the message very hard to read. Can you wrap it in triple-backticks instead (like I did with your original message)? Thanks!

**Comment by marsam** [comment]
I'm not familiar with windows but seems, but seems that`flycheck-r-has-lintr` is not quoting correctly, can you try changing it to:

```diff
diff --git a/flycheck.el b/flycheck.el
index 2b2a431..4228cae 100644
--- a/flycheck.el
+++ b/flycheck.el
@@ -8824,7 +8824,7 @@ expression, which selects linters for lintr."
     (let ((process-environment (append '("LC_ALL=C") process-environment)))
       (call-process R nil t nil
                     "--slave" "--restore" "--no-save" "-e"
-                    "library('lintr')")
+                    (shell-quote-argument "library('lintr')"))
       (goto-char (point-min))
       (not (re-search-forward "there is no package called 'lintr'" nil 'no-error)))))

```

**Comment by cpitclaudel** [comment]
Hmm. Why would you need shell-quoting with call-process?

**Comment by marsam** [comment]
It was just a rough guess, because the error message show that it considers the command upto the first single quote.  I don't know how `call-process` is internally implemented, sorry if does not makes sense.

Can try removing the single quote from `flycheck-r-has-lintr`:
    
```diff
diff --git a/flycheck.el b/flycheck.el
index 2b2a431..1cd02db 100644
--- a/flycheck.el
+++ b/flycheck.el
@@ -8824,7 +8824,7 @@ expression, which selects linters for lintr."
     (let ((process-environment (append '("LC_ALL=C") process-environment)))
       (call-process R nil t nil
                     "--slave" "--restore" "--no-save" "-e"
-                    "library('lintr')")
+                    "library(lintr)")
       (goto-char (point-min))
       (not (re-search-forward "there is no package called 'lintr'" nil 'no-error)))))
```

**Comment by arnyeinstein** [comment]
```
Suspicious state from syntax checker r-lintr: Flycheck checker r-lintr returned non-zero exit code 1, but its output contained no errors: 'c:\PROGRA~1\R\R-34~1.3/bin/x64/Rterm.exe" --slave --restore --no-save -e "library' is not recognized as an internal or external command,
operable program or batch file.
```

**Comment by marsam** [comment]
I've setup a windows machine, and confirmed this issue, but I'm not sure about the underlying problem, looks like `R.exe` misquotes the arguments.

I'm not sure how to solve this, we are basically using the same as the [syntastic.vim command](https://github.com/jimhester/lintr/blob/9c92345c4a4444b1d3be2ce4988ccc649656dace/inst/syntastic/lintr.vim#L59-L62), which also uses the [R executable](https://github.com/jimhester/lintr/blob/9c92345c4a4444b1d3be2ce4988ccc649656dace/inst/syntastic/lintr.vim#L80)

I the meantime, you can either uninstall `rlintr` or use `Rscript.exe`:

```diff
diff --git a/flycheck.el b/flycheck.el
index 2b2a431..d48db5d 100644
--- a/flycheck.el
+++ b/flycheck.el
@@ -8832,14 +8832,14 @@ expression, which selects linters for lintr."
   "An R style and syntax checker using the lintr package.

 See URL `https://github.com/jimhester/lintr'."
-  :command ("R" "--slave" "--restore" "--no-save" "-e"
+  :command ("Rscript" "--slave" "--restore" "--no-save" "-e"
             (eval (concat
                    "library(lintr);"
                    "try(lint(commandArgs(TRUE)"
                    ", cache=" (if flycheck-lintr-caching "TRUE" "FALSE")
                    ", " flycheck-lintr-linters
                    "))"))
-            "--args" source)
+            source)
   :error-patterns
   ((info line-start (file-name) ":" line ":" column ": style: " (message)
          line-end)
```

You might also need to set the `Rscript` path:

```elisp
(setq flycheck-r-lintr-executable "C:\\Program Files\\R\\R-3.4.3\\bin\\x64\\Rscript.exe")
```


Tested with:

-   Windows 7
-   R version 3.4.3(2017-11-30) &#x2013; latest stable
-   Emacs 2.5.5 &#x2013; latest stable

**Comment by arnyeinstein** [comment]
Thanks Marsam, the changes (including the path) did solve the issue.
Renger

**Comment by fmdkdd** [comment]
@yiufung We are certainly not changing the default executable to `Rscript.exe`, as that would be a Windows-only fix.  Besides, it's easy enough to change in your config.

The shell quoting issue has been fixed, as far as I can tell.

**Comment by yiufung** [comment]
@fmdkdd You're right, certainly we're not replacing it with the `.exe` ending. The only way I can think of to make my config portable without modifying `flycheck.el` every time in a new machine, is to redefine `r-lintr` after loading flycheck: 

```emacs-lisp
(flycheck-define-checker r-lintr
  "An R style and syntax checker using the lintr package.

See URL `https://github.com/jimhester/lintr'."
  :command ("RScript" "--slave" "--restore" "--no-save" "-e"
            (eval (concat
                   "library(lintr);"
                   "try(lint(commandArgs(TRUE)"
                   ", cache=" (if flycheck-lintr-caching "TRUE" "FALSE")
                   ", " flycheck-lintr-linters
                   "))"))
            source)
  :error-patterns
  ((info line-start (file-name) ":" line ":" column ": style: " (message)
         line-end)
   (warning line-start (file-name) ":" line ":" column ": warning: " (message)
            line-end)
   (error line-start (file-name) ":" line ":" column ": error: " (message)
          line-end))
  :modes ess-mode
  :predicate
  ;; Don't check ESS files which do not contain R, and make sure that lintr is
  ;; actually available
  (lambda ()
    (and (equal ess-language "S")
         (flycheck-r-has-lintr (flycheck-checker-executable 'r-lintr))))
  :verify (lambda (checker)
            (let ((has-lintr (flycheck-r-has-lintr
                              (flycheck-checker-executable checker))))
              (list
               (flycheck-verification-result-new
                :label "lintr library"
                :message (if has-lintr "present" "missing")
                :face (if has-lintr 'success '(bold error)))))))
```
which is huge. 

Meanwhile, I tried with the two commands under Windows and my Debian VM: 

`Rscript --slave --restore --no-save -e "library('methods'); print('hello')"`
`R --slave --restore --no-save -e "library('methods'); print('hello')"`

Debian works well on both cases, as well as Windows `RScript`, but Windows `R` gives the familiar error: 

`'C:\PROGRA~1\R\R-34~1.3/bin/x64/Rterm.exe" --slave --restore --no-save -e "library' is not recognized as an internal or external command, operable program or batch file.`

Note how it strangely changed to Rterm instead. I wonder if it has anything to do with: 

 - From [Invoking R from command line](https://cran.r-project.org/doc/manuals/R-intro.html#Invoking-R-from-the-command-line), it says `R` with `-e expression` takes expression, and needs to be quoted. 
- From [Scripting with R](https://cran.r-project.org/doc/manuals/R-intro.html#Scripting-with-R), it says `Short scripts can be passed to Rscript on the command-line via the -e flag. (Empty scripts are not accepted.) `, it seems this is more a literal string script rather than expression. 

It seems `RScript` is a little more preferred way to invoke script, from the docs, from usage frequency in R source code, and some [discussion](https://stackoverflow.com/questions/3412911/r-exe-rcmd-exe-rscript-exe-and-rterm-exe-whats-the-difference), though I haven't tested it fully in flycheck yet.

**Comment by fmdkdd** [comment]
> which is huge.

Couldn't you just replace the `:command` property with `(setf (flycheck-checker-get 'r-lintr 'command) '("RScript" ...))`?

> It seems RScript is a little more preferred way to invoke script,

If `Rscript ` is indeed portable, fixes the Windows bug while not impairing functionality for other OSes, then we could switch to it, yes.  I don't use R, so I really couldn't answer those questions.

cc @marsam

**Comment by yiufung** [comment]
@fmdkdd Thanks for the advice, it's very useful. I'm just starting on elisp so that doesn't occur to me. 

I was putting it under `:config` in `use-package`, but I saw this on startup: `Debugger entered--Lisp error: (void-function \(setf\ flycheck-checker-get\))`. I `defun`ed it and added to `flycheck-mode-hook` and now it works. Thanks.
